pytorch_lightning/core/lightning.py

```diff
@@ -91,11 +91,42 @@ def forward(self, x):
         if self.trainer.proc_rank == 0:
             print(*args, **kwargs)
 
+    def __call__(self, *data, **kwargs):
+        r"""
+        Automatically moves data to correct device if possible, then call torch.nn.Module.__call__
+        Lightning will warn you if it automatically moves any data
+
+        Args:
+            *data: Any positional arguments for torch.nn.Module.__call__. These are typically input data
+            **kwargs: Any keyword arguments for torch.nn.Module.__call__
+
+        Example:
+
+            .. code-block:: python
+
+                model = model.cuda(0)
+                model.prepare_data()
+                loader = model.train_dataloader()
+                for x, y in loader:
+                    output = model(x)  # Lightning will automove data here and warn you of it
+
+        """
+        devices = [p.device for p in self.parameters()]
+        # All parameters must be on same device to automove data
+        # Otherwise we just do what nn.Module does normally
+        if len(set(devices)) == 1:
+            device = devices[0]
+            data = transfer_data_to_device(data, device.type, device.index, warn_on_transfer=True)
+            kwargs = transfer_data_to_device(kwargs, device.type, device.index, warn_on_transfer=True)
+        return super(LightningModule, self).__call__(*data, **kwargs)
```

Comment:
@Borda  and I discussed this and we both agree, that we shouldn't do this in the Module (at least not by default). In our opinion we should always be able to use lightning module as nn module.

What I propose is the following:

We change this part to a decorator, that can be added to forward and is automatically added from trainer. (sorry for that coming so late).


In that case you could make the decorator a class, that also caches devices eventually :)

Reply:
Love the idea of adding the decorator dynamically there!